File: Algorithms/0695_Max_Area_of_Island/Java/MaxAreaofIslandSolution2.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int res = 0;
        for (int y = 0; y < m; y++) {
            for (int x = 0; x < n; x++) {
                if (grid[y][x] == 1) {
                    Stack<int[]> stack = new Stack<>();
                    int tempRes = 0;
                    grid[y][x] = 0;
                    stack.add(new int[]{y, x});
                    while (!stack.empty()) {
                        int[] cur = stack.pop();
                        tempRes += 1;
                        if (0 <= cur[0] - 1 && cur[0] - 1 < m && 0 <= cur[1] && cur[1] < n && grid[cur[0] - 1][cur[1]] == 1) {
                            grid[cur[0] - 1][cur[1]] = 0;
                            stack.add(new int[]{cur[0] - 1, cur[1]});
                        }
                        if (0 <= cur[0] + 1 && cur[0] + 1 < m && 0 <= cur[1] && cur[1] < n && grid[cur[0] + 1][cur[1]] == 1) {
                            grid[cur[0] + 1][cur[1]] = 0;
                            stack.add(new int[]{cur[0] + 1, cur[1]});
                        }
                        if (0 <= cur[0] && cur[0] < m && 0 <= cur[1] - 1 && cur[1] - 1 < n && grid[cur[0]][cur[1] - 1] == 1) {
                            grid[cur[0]][cur[1] - 1] = 0;
                            stack.add(new int[]{cur[0], cur[1] - 1});
                        }
                        if (0 <= cur[0] && cur[0] < m && 0 <= cur[1] + 1 && cur[1] + 1 < n && grid[cur[0]][cur[1] + 1] == 1) {
                            grid[cur[0]][cur[1] + 1] = 0;
                            stack.add(new int[]{cur[0], cur[1] + 1});
                        }
                    }
                    res = Math.max(res, tempRes);
                }
            }
        }
        return res;
    }
}

File: Algorithms/0695_Max_Area_of_Island/Python/Max_Area_of_Island_Solution_2.py

# Space: O(n)
# Time: O(n)

class Solution:
    def maxAreaOfIsland(self, grid):
        res = 0
        m = len(grid)
        n = len(grid[0])

        for y in range(m):
            for x in range(n):
                if grid[y][x] == 1:
                    temp_res = 0
                    grid[y][x] = 0
                    stack = [[y, x]]
                    while stack:
                        cur_y, cur_x = stack.pop()
                        temp_res += 1
                        if 0 <= cur_y - 1 < m and 0 <= cur_x < n and grid[cur_y - 1][cur_x] == 1:
                            grid[cur_y - 1][cur_x] = 0
                            stack.append([cur_y - 1, cur_x])
                        if 0 <= cur_y + 1 < m and 0 <= cur_x < n and grid[cur_y + 1][cur_x] == 1:
                            grid[cur_y + 1][cur_x] = 0
                            stack.append([cur_y + 1, cur_x])
                        if 0 <= cur_y < m and 0 <= cur_x - 1 < n and grid[cur_y][cur_x - 1] == 1:
                            grid[cur_y][cur_x - 1] = 0
                            stack.append([cur_y, cur_x - 1])
                        if 0 <= cur_y < m and 0 <= cur_x + 1 < n and grid[cur_y][cur_x + 1] == 1:
                            grid[cur_y][cur_x + 1] = 0
                            stack.append([cur_y, cur_x + 1])
                    res = max(res, temp_res)
        return res





